src/EventListener/ProductTranslationUpdateListener.php

```diff
@@ -0,0 +1,139 @@
+                                ValidatorInterface $validator,
+                                FlashBagInterface $flashBag
+    )
+    {
+        $this->request = $requestStack->getCurrentRequest();
+        $this->redirectionFactory = $redirectionFactory;
+        $this->redirectionRepository = $redirectionRepository;
+        $this->infiniteLoopResolver = $infiniteLoopResolver;
+        $this->router = $router;
+        $this->validator = $validator;
+        $this->flashBag = $flashBag;
+    }
+
+    public function postUpdate(LifecycleEventArgs $event): void
+    {
+        $productTranslation = $event->getObject();
+        if (!$productTranslation instanceof ProductTranslationInterface) {
+            return;
+        }
+
+        $uow = $event->getEntityManager()->getUnitOfWork();
+        $changeSet = $uow->getEntityChangeSet($productTranslation);
+        $this->handleAutomaticRedirectionCreation($productTranslation, $changeSet);
+    }
+
+    private function handleAutomaticRedirectionCreation(ProductTranslationInterface $productTranslation, $changeSet): void
+    {
+        if (empty($changeSet['slug'])) {
+            return;
+        }
+
+        if (null === $this->request) {
+            return;
+        }
+
+        $postProductParams = $this->request->request->get('sylius_product', []);
+        if (empty($postProductParams)) {
+            return;
+        }
+        $localeCode = $productTranslation->getLocale();
+
+        if (empty($postProductParams['translations'][$localeCode])) {
+            return;
+        }
+
+        $postProductTranslation = $postProductParams['translations'][$localeCode];
+        if (empty($postProductTranslation['addAutomaticRedirection'])) {
+            return;
+        }
+
+        $oldSlug = $changeSet['slug'][0];
+        $newSlug = $changeSet['slug'][1];
+
+        /** @var RedirectInterface $redirect */
+        $redirect = $this->redirectionFactory->createNew();
+        $source = $this->router->generate('sylius_shop_product_show', ['slug' => $oldSlug]);
+        $destination = $this->router->generate('sylius_shop_product_show', ['slug' => $newSlug]);
+
+        $redirect->setSource($source);
+        $redirect->setDestination($destination);
+        $redirect->enable();
+        $redirect->setOnly404(false);
+        $redirect->setPermanent(true);
+
+        $conflictingRedirect = $this->infiniteLoopResolver->getConflictingRedirect($redirect);
+        if ($conflictingRedirect instanceof RedirectInterface) {
+            $this->redirectionRepository->remove($conflictingRedirect);
+        }
+
+        $violations = $this->validator->validate($redirect, null, ['sylius']);
+        if ($violations->count() > 0) {
+            /** @var ConstraintViolationInterface $violation */
+            foreach ($violations as $violation) {
+                $this->flashBag->add('error', [
+                    'message' => $violation->getMessageTemplate(),
```

Comment:
This message should go inside the `UpdateHandlingException` I suppose

Reply:
No, if I set it inside the `UpdateHandlingException` it automatically adds a `sylius.resource.` to the message, and it's quite weird to re-translate all this

Reply:
The problem with this approach is that it gives two flash messages I suppose? One with a generic error message and then the one you are giving. Isn't that what is happening?

You should be able to add only the translation we need to the `sylius.resource` namespace.

Reply:
There is still a problem. You can't pass translation parameters through `UpdateHandlingException`

Reply:
Okay, I understand :D